### PR TITLE
Llama: SDPA FA2 path + static cache fix

### LINKED ISSUE #30417: Bad generation results with Llama after v4.39.0
### System Info

transformers version: 4.41.0.dev0
Platform: Linux-5.15.0-89-generic-x86_64-with-glibc2.
Python version: 3.10.
Huggingface_hub version: 0.20.
Safetensors version: 0.4.
Accelerate version: 0.21.0

### Who can help?

@ArthurZucker @gante

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

The generation output quality with the current 4.41.0.dev0 version is very bad compared to the previous 4.39.0 version, at least with Llama. With quantized models, it outputs complete gibberish. The same code works totally fine with 4.39.0

```Python
import torch, os
from transformers import AutoModelForCausalLM, AutoTokenizer

model_id  = "meta-llama/Llama-2-7b-chat-hf"
model     = AutoModelForCausalLM.from_pretrained(model_id, cache_dir='.', torch_dtype=torch.float16, attn_implementation="sdpa").to('cuda').eval();
tokenizer = AutoTokenizer.from_pretrained(model_id, cache_dir='.') 

tokenizer.add_bos_token = False
tokenizer.add_eos_token = False

prompt = "<s>[INST] How do I build a car? [/INST]"

gen_out = model.generate(**tokenizer([prompt], return_tensors="pt").to(model.device), do_sample=False, 
                                                cache_implementation="static", max_new_tokens=100, pad_token_id=tokenizer.eos_token_id, 
                                                temperature=None, top_p=None, use_cache=False)

print()
print(tokenizer.decode(gen_out[0]))
```

```
# version: 4.39.0 - works as expected
<s> [INST] How do I build a car? [/INST]  Building a car is a complex and challenging project that requires a significant amount of time, money, and expertise. Here are some general steps that you might consider when building a car:

1. Define your goals: What kind of car do you want to build? What features do you want to include? What is your budget? Answering these questions will help you determine the scope of your project and what you need to do to get started.
2. Research and plan:
```

```
# version: 4.41.0 - bad output, outputs gibberish 
<s> [INST] How do I build a car? [/INST]  I's (the 0-2) are dots d's the traveling 4 v5 8 out the9 of the9 1t 1 ch do not always and the9 10 11 is-not 1 rt 1 c0 the0.

To build a car, you will need to have a good understanding of mechanical systems, electrical systems, and fabrication techniques. You will also need to have a
```

### Expected behavior

The output should be the same as with the previous 4.39.0 version

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

## Problem
The recently enabled SDPA FA2 path doesn't pass the `attn_mask` (causal mask) argument. As such, when the static cache is used, `S` (see the [SDPA docs](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) for terminology) is the full cache length as opposed to the sequence length. Therefore, the inferred mask in SDPA is incorrect, resulting in bad numerical values. 

PR that introduced the issue: https://github.com/huggingface/transformers/pull/30317. The issue was not caught in our llama testing suite because we didn't have a test for the static cache WITHOUT compilation:
- dynamic cache -> no issue with the length
- static cache + compile -> we manually build the causal attention mask, passing it to sdpa
- static cache -> we don't build the causal mask = `causal mask is None` = problem described above triggered

## Solution
There were two possible paths here:
1. Build and pass a full attention mask, corresponding to the full cache length, to avoid `causal mask is None` (#30414 );
2. Crop empty KV entries (corresponding to the empty cache) before SDPA.

I went with `2`, as it saves us tons of masked computations :) I've also ensured the static cache without compilation is numerically tested in the llama test file.

Fixes #30417 

________________________________

Slow tests ran locally: `llama`, `gemma`, `cohere`, `test_cache_utils.py`.

### CODE REVIEW

**Comment by gante** [comment]
cc @younesbelkada @fxmarty (SDPA + FA2 changes)
cc @zucchini-nlp (alternative fix to https://github.com/huggingface/transformers/pull/30414, see possible solutions in the PR header)

File: src/transformers/models/cohere/modeling_cohere.py [hunk 1]
```diff
@@ -533,10 +533,13 @@ def forward(
         cache_position: Optional[torch.LongTensor] = None,
     ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
         if output_attentions:
-            # TODO: Improve this warning with e.g. `model.config.attn_implementation = "manual"` once this is implemented.
+            # TODO: Improve this warning with e.g. `model.config.attn_implementation = "manual"` once this is
```
Comment (gante) [inline src/transformers/models/cohere/modeling_cohere.py:536]: 120 char limit OCD :D

**Review by younesbelkada — APPROVED** [review]
LGTM ! Thanks !

**Review by ArthurZucker — COMMENTED** [review]
Okay very interesting and partially related to  #30442.
Before merging, let's test that this does not affect compiled performances, as indexing can be costly.

**Comment by gante** [comment]
As Arthur wrote -- the current state of the PR adds a slowdown on the eager path. 

I'm exploring an alternative path: first standardize our static cache to behave like our other caches (living outside the model as a stand-alone object), then forcing the generation of the full mask if a static cache is used (which fixes this issue).

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -1073,6 +1080,7 @@ def _update_causal_mask(
         if self.config._attn_implementation == "sdpa":
             # For SDPA, when possible, we will rely on its `is_causal` argument instead of its `attn_mask` argument,
             # in order to dispatch on Flash Attention 2.
+            breakpoint()
```
Comment (fxmarty) [inline src/transformers/models/llama/modeling_llama.py:1083]: remove

File: tests/models/llama/test_modeling_llama.py [hunk 3]
```diff
@@ -684,17 +683,18 @@ def test_model_13b_greedy_generation(self):
     @require_torch_gpu
     @require_read_token
     def test_compile_static_cache(self):
+        # `torch==2.2` will throw an error on this test (as in other compilation tests), but torch==2.1.2 and torch>2.2
+        # work as intended. See https://github.com/pytorch/pytorch/issues/121943
```
Comment (fxmarty) [inline tests/models/llama/test_modeling_llama.py:687]: you can skipTest on torch version

File: src/transformers/models/cohere/modeling_cohere.py (hunk 1 above)
Comment (fxmarty) [reply src/transformers/models/cohere/modeling_cohere.py:536]: It is not in `make style` is it? https://github.com/huggingface/transformers/blob/a98c41798cf6ed99e1ff17e3792d6e06a2ff2ff3/pyproject.toml#L5

File: tests/models/llama/test_modeling_llama.py [hunk 4]
```diff
@@ -706,38 +706,25 @@ def test_compile_static_cache(self):
...
+        static_text = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)
+        self.assertEqual(EXPECTED_TEXT_COMPLETION, static_text)
+
+        # Static Cache + compile
+        model.forward = torch.compile(model.forward, mode="reduce-overhead", fullgraph=True)
+        generated_ids = model.generate(
+            **inputs, max_new_tokens=NUM_TOKENS_TO_GENERATE, do_sample=False, cache_implementation="static"
+        )
+        static_compiled_text = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)
+        self.assertEqual(EXPECTED_TEXT_COMPLETION, static_compiled_text)
```
Comment (fxmarty) [inline tests/models/llama/test_modeling_llama.py:727]: Shouldn't the static cache tests be somewhere like test_modeling_common.py and test every supported models?

**Review by fxmarty — COMMENTED** [review]
Thanks a lot for noticing, indeed I should have tested static without compile. It is very good to add a test for it.

I am wondering - couldn this change cause issues with cuda graph capture, adding dynamicity in the tensor shapes? Making the capture slower?

File: tests/models/llama/test_modeling_llama.py [hunk 5]
```diff
@@ -684,17 +683,18 @@ def test_model_13b_greedy_generation(self):
     @require_torch_gpu
     @require_read_token
     def test_compile_static_cache(self):
+        # `torch==2.2` will throw an error on this test (as in other compilation tests), but torch==2.1.2 and torch>2.2
+        # work as intended. See https://github.com/pytorch/pytorch/issues/121943
         NUM_TOKENS_TO_GENERATE = 40
-        EXPECTED_TEXT_COMPLETION = {
-            7: [
```
Comment (younesbelkada) [inline tests/models/llama/test_modeling_llama.py:689]: note you might get different results from A10 to T4, hence this dict. this change can lead to the push-important-models test + Slow tests to fail 😢 
You can now SSH into our runners and get the value of the generations for each device type

**Comment by gante** [comment]
(closed in favor of #30476, a much better long-term solution)
